<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Комментарий ревьюера</b>


Привет, Денис! Мы уже встречались на предыдущем проекте, но представлюсь повторно! Меня зовут Дмитрий Махортов, и я буду проверять твой проект. Сразу предлагаю общение на «ты» 🙂, но если тебе это не комфортно, то дай знать, и мы перейдем на «вы».


Моя основная цель — не указать на совершенные тобою ошибки, а поделиться своим опытом и помочь тебе погрузиться в увлекательный мир работы с данными и вырасти в крепкого профи. Это отдаленная цель. А ближайшая - сделать твою работу еще лучше )).
   
    
Все ключевые этапы в работе выполнены, и я вижу что с проектом ты справшяешься. Есть моменты, которые нужно доработать, но я уверен, у тебя все получится.
        
Расскажу как обычно проходит проверка:    
Бывают моменты, которые требуют пристального внимания. Комментарии по ним выделены <span style='background-color:#F7B3A4'> красным цветом </span> и обозначены значком 🛑. После их доработки проект будет принят.  🙂

<span style='background-color:#B7EBA7'> Зеленым цветом </span> и значком ✅ отмечены удачные и элегантные решения, на которые можно опираться в будущих проектах. Или советы «со звездочкой», которые помогут тебе в будущем.

<span style='background-color:#F9EDA6'>Жёлтым цветом </span> и значком ⚠️ выделено то, что в следующий раз можно сделать по-другому. Ты можешь учесть эти комментарии при выполнении будущих заданий или доработать проект сейчас (однако это не обязательно).

Давай работать над проектом в диалоге: **если ты что-то меняешь в проекте по моим рекомендациям — пиши об этом**. Выбери для своих комментариев какой-то заметный цвет, так мне будет легче отследить изменения. Пожалуйста, **не перемещай, не изменяй и не удаляй мои комментарии**. Всё это поможет выполнить повторную проверку твоего проекта оперативнее.
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента</b></font>
   
Привет!

</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>образец комментария студента</b></font>
   
Можешь использовать этот комментарий. Просто удали этот текст и вместо него напиши свой 🙂

</div>

# Прекод

# Сборный проект-4

Вам поручено разработать демонстрационную версию поиска изображений по запросу.

Для демонстрационной версии нужно обучить модель, которая получит векторное представление изображения, векторное представление текста, а на выходе выдаст число от 0 до 1 — покажет, насколько текст и картинка подходят друг другу.

### Описание данных

Данные доступны по [ссылке](https://code.s3.yandex.net/datasets/dsplus_integrated_project_4.zip).

В файле `train_dataset.csv` находится информация, необходимая для обучения: имя файла изображения, идентификатор описания и текст описания. Для одной картинки может быть доступно до 5 описаний. Идентификатор описания имеет формат `<имя файла изображения>#<порядковый номер описания>`.

В папке `train_images` содержатся изображения для тренировки модели.

В файле `CrowdAnnotations.tsv` — данные по соответствию изображения и описания, полученные с помощью краудсорсинга. Номера колонок и соответствующий тип данных:

1. Имя файла изображения.
2. Идентификатор описания.
3. Доля людей, подтвердивших, что описание соответствует изображению.
4. Количество человек, подтвердивших, что описание соответствует изображению.
5. Количество человек, подтвердивших, что описание не соответствует изображению.

В файле `ExpertAnnotations.tsv` содержатся данные по соответствию изображения и описания, полученные в результате опроса экспертов. Номера колонок и соответствующий тип данных:

1. Имя файла изображения.
2. Идентификатор описания.

3, 4, 5 — оценки трёх экспертов.

Эксперты ставят оценки по шкале от 1 до 4, где 1 — изображение и запрос совершенно не соответствуют друг другу, 2 — запрос содержит элементы описания изображения, но в целом запрос тексту не соответствует, 3 — запрос и текст соответствуют с точностью до некоторых деталей, 4 — запрос и текст соответствуют полностью.

В файле `test_queries.csv` находится информация, необходимая для тестирования: идентификатор запроса, текст запроса и релевантное изображение. Для одной картинки может быть доступно до 5 описаний. Идентификатор описания имеет формат `<имя файла изображения>#<порядковый номер описания>`.

В папке `test_images` содержатся изображения для тестирования модели.

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 👍 Да, это хорошая практика - дать описание контекста и проблемы, которую мы решаем. </div>

## 1. Исследовательский анализ данных

Наш датасет содержит экспертные и краудсорсинговые оценки соответствия текста и изображения.

В файле с экспертными мнениями для каждой пары изображение-текст имеются оценки от трёх специалистов. Для решения задачи вы должны эти оценки агрегировать — превратить в одну. Существует несколько способов агрегации оценок, самый простой — голосование большинства: за какую оценку проголосовала большая часть экспертов (в нашем случае 2 или 3), та оценка и ставится как итоговая. Поскольку число экспертов меньше числа классов, может случиться, что каждый эксперт поставит разные оценки, например: 1, 4, 2. В таком случае данную пару изображение-текст можно исключить из датасета.

Вы можете воспользоваться другим методом агрегации оценок или придумать свой.

В файле с краудсорсинговыми оценками информация расположена в таком порядке:

1. Доля исполнителей, подтвердивших, что текст **соответствует** картинке.
2. Количество исполнителей, подтвердивших, что текст **соответствует** картинке.
3. Количество исполнителей, подтвердивших, что текст **не соответствует** картинке.

После анализа экспертных и краудсорсинговых оценок выберите либо одну из них, либо объедините их в одну по какому-то критерию: например, оценка эксперта принимается с коэффициентом 0.6, а крауда — с коэффициентом 0.4.

Ваша модель должна возвращать на выходе вероятность соответствия изображения тексту, поэтому целевая переменная должна иметь значения от 0 до 1.


In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.linear_model import LogisticRegression

from tensorflow.keras.preprocessing.image import ImageDataGenerator

from scipy.stats import mode

import torch
import torchvision.models as models
import torchvision.transforms as transforms
from PIL import Image
import os
from tqdm import tqdm

from transformers import BertTokenizer, BertModel

from sklearn.model_selection import GroupShuffleSplit

from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_absolute_error, r2_score
from sklearn.pipeline import make_pipeline
from sklearn.model_selection import GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.neural_network import MLPRegressor
from sklearn.utils import resample

import matplotlib.image as mpimg
from sklearn.metrics.pairwise import cosine_similarity
import torch.nn as nn


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично, все нужные библиотеки импортированы в начале ноутбука.Это хорошая практика.</div>

In [2]:
import warnings
warnings.filterwarnings('ignore')

In [3]:
def data_exploration_func(df_list):
    for df_current in df_list:
        print(f'Изучение данных датафрейма "{df_current.name}"\n')
        print(f'Первые 10 строк "{df_current.name}"')
        display(df_current.head(10))
        print('\n')
        print(f'"{df_current.name}" состоит из:')
        print(f'{df_current.shape[0]} строк')
        print(f'{df_current.shape[1]} столбцов')
        print('\n')
        print(f'Общая информация о "{df_current.name}"\n')
        df_current.info()
        print('\n')
        print(f'Описательная статистика "{df_current.name}"')
        display(df_current.describe(include='all').T)
        print('\n')

In [4]:
RANDOM_STATE=42

In [5]:
!wget -q https://code.s3.yandex.net/datasets/dsplus_integrated_project_4.zip
!unzip -q dsplus_integrated_project_4.zip -d /content
DATA_PATH = '/content/to_upload'
df_train = pd.read_csv(os.path.join(DATA_PATH, 'train_dataset.csv'), sep=',')
df_crowdAnnotations = pd.read_csv(
    os.path.join(DATA_PATH, 'CrowdAnnotations.tsv'),
    sep='\t',
    names=['image', 'query_id', 'desc_perc', 'n_match', 'n_no_match'],
    index_col=False
)
df_expertAnnotations = pd.read_csv(
    os.path.join(DATA_PATH, 'ExpertAnnotations.tsv'),
    sep='\t',
    names=['image', 'query_id', 'ex_1', 'ex_2', 'ex_3'],
    index_col=False
)
df_test = pd.read_csv(os.path.join(DATA_PATH, 'test_queries.csv'), sep='|')
image_folder = os.path.join(DATA_PATH, 'train_images')
test_images = os.path.join(DATA_PATH, 'test_images')

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Очень "больная тема" - это загрузка данных и использование путей.Очевидно, что на моей машине (и скорее всего машине любого другого пользователя) пути к данным будут другие. И это "рабочая ситуация для большинства проектов. Решается она следующим образом:

    
  - Путь к директории с данными  должен быть задан в константе, в самом начале работы. Тогда ревьюеру (или любому другому пользователю твоего ноутбука, даже тебе самому, если нужно будет провести презентацию руководителю на другом компьютере) нужно будет изменить только этот путь.
  - Все остальные пути внутри ноутбука должны быть созданы с исопользованием этой константы, а не заданы жестко (есть сленговое определение "хардкод").  Более  того, настоятельно рекомендую для создания сложных путей вместо простой конкатерации строк с исползованием '/' использовать os.path.join(), тогда код будет одинаково надежно работать и в среде Windows и Linux (MacOS).
   
Чтобы запустить твой код при проверке я был вынужден руками менять код по всему ноутбуку. Прошу тебя внести изменения так, чтобы путь к данным нужно было изменять всего один раз.


А вот это образец универсального кода для загрузки данных    
    
    DATA_PATH = твой путь    
    
    df_train = pd.read_csv(os.path.join(DATA_PATH, 'train_dataset.csv'))
    ....
    
    
</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
    <b>v2</b> исправлено 👍 </div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
На мой взгляд привязка к локальному диску (или твоему личному гугл драйву) это не очень  удобно, т.к. у того, кто попытается воспроизвсти твой проект свой диск и нет доступа к твоему датасету. Выходом может быть прямое закачивание датасета в сессию colab. Вот код для скачивания и извлечения данных из архива
    
    
    # загрузка датасета

    !wget -q https://code.s3.yandex.net/datasets/dsplus_integrated_project_4.zip
    !unzip -q dsplus_integrated_project_4.zip

    DATA_PATH = '/content/to_upload'
    
----------------
    
Но справедливости ради хочу заметить, что для личных проектов использование связки Google Drive + Colab очень удобно,т.к. позволяет сохранять на "безопасный" диск не только датасеты,но и промежуточные результаты (эмбеддинги, данные для обучения, промежуточные веса моделей, если вдруг модель обучается очень долго).    
</div>

In [6]:
df_train.name= 'df_train'
df_crowdAnnotations.name = 'df_crowdAnnotations'
df_expertAnnotations.name = 'df_expertAnnotations'
df_test.name = 'df_test'

In [7]:
df_list = [df_train, df_crowdAnnotations, df_expertAnnotations, df_test]
data_exploration_func(df_list)

Изучение данных датафрейма "df_train"

Первые 10 строк "df_train"


image                     query_id  \
0  1056338697_4f7d7ce270.jpg  2549968784_39bfbe44f9.jpg#2   
1  1262583859_653f1469a9.jpg  2549968784_39bfbe44f9.jpg#2   
2  2447284966_d6bbdb4b6e.jpg  2549968784_39bfbe44f9.jpg#2   
3  2549968784_39bfbe44f9.jpg  2549968784_39bfbe44f9.jpg#2   
4  2621415349_ef1a7e73be.jpg  2549968784_39bfbe44f9.jpg#2   
5  3030566410_393c36a6c5.jpg  2549968784_39bfbe44f9.jpg#2   
6  3155451946_c0862c70cb.jpg  2549968784_39bfbe44f9.jpg#2   
7  3222041930_f642f49d28.jpg  2549968784_39bfbe44f9.jpg#2   
8   343218198_1ca90e0734.jpg  2549968784_39bfbe44f9.jpg#2   
9  3718964174_cb2dc1615e.jpg  2549968784_39bfbe44f9.jpg#2   

                                          query_text  
0  A young child is wearing blue goggles and sitt...  
1  A young child is wearing blue goggles and sitt...  
2  A young child is wearing blue goggles and sitt...  
3  A young child is wearing blue goggles and sitt...  
4  A young child is wearing blue goggles and sitt...  
5  A young child is wearing blue goggles and sitt...  
6  A young child is wearing blue goggles and sitt...  
7  A young child is wearing blue goggles and sitt...  
8  A young child is wearing blue goggles and sitt...  
9  A young child is wearing blue goggles and sitt...



"df_train" состоит из:
5822 строк
3 столбцов


Общая информация о "df_train"

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5822 entries, 0 to 5821
Data columns (total 3 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   image       5822 non-null   object
 1   query_id    5822 non-null   object
 2   query_text  5822 non-null   object
dtypes: object(3)
memory usage: 136.6+ KB


Описательная статистика "df_train"


count unique                                                top  \
image       5822   1000                           754852108_72f80d421f.jpg   
query_id    5822    977                        2600867924_cd502fc911.jpg#2   
query_text  5822    977  Two dogs , one brown and white and one black a...   

           freq  
image        10  
query_id     34  
query_text   34



Изучение данных датафрейма "df_crowdAnnotations"

Первые 10 строк "df_crowdAnnotations"


image                     query_id  desc_perc  n_match  \
0  1056338697_4f7d7ce270.jpg  1056338697_4f7d7ce270.jpg#2   1.000000        3   
1  1056338697_4f7d7ce270.jpg   114051287_dd85625a04.jpg#2   0.000000        0   
2  1056338697_4f7d7ce270.jpg  1427391496_ea512cbe7f.jpg#2   0.000000        0   
3  1056338697_4f7d7ce270.jpg  2073964624_52da3a0fc4.jpg#2   0.000000        0   
4  1056338697_4f7d7ce270.jpg  2083434441_a93bc6306b.jpg#2   0.000000        0   
5  1056338697_4f7d7ce270.jpg  2204550058_2707d92338.jpg#2   0.000000        0   
6  1056338697_4f7d7ce270.jpg  2224450291_4c133fabe8.jpg#2   0.000000        0   
7  1056338697_4f7d7ce270.jpg  2248487950_c62d0c81a9.jpg#2   0.333333        1   
8  1056338697_4f7d7ce270.jpg  2307118114_c258e3a47e.jpg#2   0.000000        0   
9  1056338697_4f7d7ce270.jpg  2309860995_c2e2a0feeb.jpg#2   0.000000        0   

   n_no_match  
0           0  
1           3  
2           3  
3           3  
4           3  
5           3  
6           3  
7           2  
8           3  
9           3



"df_crowdAnnotations" состоит из:
47830 строк
5 столбцов


Общая информация о "df_crowdAnnotations"

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 47830 entries, 0 to 47829
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   image       47830 non-null  object 
 1   query_id    47830 non-null  object 
 2   desc_perc   47830 non-null  float64
 3   n_match     47830 non-null  int64  
 4   n_no_match  47830 non-null  int64  
dtypes: float64(1), int64(2), object(2)
memory usage: 1.8+ MB


Описательная статистика "df_crowdAnnotations"


count unique                         top freq      mean  \
image         47830   1000   1572532018_64c030c974.jpg  120       NaN   
query_id      47830   1000  249394748_2e4acfbbb5.jpg#2  169       NaN   
desc_perc   47830.0    NaN                         NaN  NaN  0.068786   
n_match     47830.0    NaN                         NaN  NaN  0.208488   
n_no_match  47830.0    NaN                         NaN  NaN  2.820155   

                 std  min  25%  50%  75%  max  
image            NaN  NaN  NaN  NaN  NaN  NaN  
query_id         NaN  NaN  NaN  NaN  NaN  NaN  
desc_perc   0.207532  0.0  0.0  0.0  0.0  1.0  
n_match     0.628898  0.0  0.0  0.0  0.0  5.0  
n_no_match  0.656676  0.0  3.0  3.0  3.0  6.0



Изучение данных датафрейма "df_expertAnnotations"

Первые 10 строк "df_expertAnnotations"


image                     query_id  ex_1  ex_2  ex_3
0  1056338697_4f7d7ce270.jpg  2549968784_39bfbe44f9.jpg#2     1     1     1
1  1056338697_4f7d7ce270.jpg  2718495608_d8533e3ac5.jpg#2     1     1     2
2  1056338697_4f7d7ce270.jpg  3181701312_70a379ab6e.jpg#2     1     1     2
3  1056338697_4f7d7ce270.jpg  3207358897_bfa61fa3c6.jpg#2     1     2     2
4  1056338697_4f7d7ce270.jpg  3286822339_5535af6b93.jpg#2     1     1     2
5  1056338697_4f7d7ce270.jpg  3360930596_1e75164ce6.jpg#2     1     1     1
6  1056338697_4f7d7ce270.jpg  3545652636_0746537307.jpg#2     1     1     1
7  1056338697_4f7d7ce270.jpg   434792818_56375e203f.jpg#2     1     1     2
8   106490881_5a2dd9b7bd.jpg  1425069308_488e5fcf9d.jpg#2     1     1     1
9   106490881_5a2dd9b7bd.jpg  1714316707_8bbaa2a2ba.jpg#2     2     2     2



"df_expertAnnotations" состоит из:
5822 строк
5 столбцов


Общая информация о "df_expertAnnotations"

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5822 entries, 0 to 5821
Data columns (total 5 columns):
 #   Column    Non-Null Count  Dtype 
---  ------    --------------  ----- 
 0   image     5822 non-null   object
 1   query_id  5822 non-null   object
 2   ex_1      5822 non-null   int64 
 3   ex_2      5822 non-null   int64 
 4   ex_3      5822 non-null   int64 
dtypes: int64(3), object(2)
memory usage: 227.6+ KB


Описательная статистика "df_expertAnnotations"


count unique                          top freq      mean       std  \
image       5822   1000    3107513635_fe8a21f148.jpg   10       NaN       NaN   
query_id    5822    977  2600867924_cd502fc911.jpg#2   34       NaN       NaN   
ex_1      5822.0    NaN                          NaN  NaN   1.43662  0.787084   
ex_2      5822.0    NaN                          NaN  NaN  1.624356  0.856222   
ex_3      5822.0    NaN                          NaN  NaN  1.881999  0.904087   

          min  25%  50%  75%  max  
image     NaN  NaN  NaN  NaN  NaN  
query_id  NaN  NaN  NaN  NaN  NaN  
ex_1      1.0  1.0  1.0  2.0  4.0  
ex_2      1.0  1.0  1.0  2.0  4.0  
ex_3      1.0  1.0  2.0  2.0  4.0



Изучение данных датафрейма "df_test"

Первые 10 строк "df_test"


Unnamed: 0                     query_id  \
0           0  1177994172_10d143cb8d.jpg#0   
1           1  1177994172_10d143cb8d.jpg#1   
2           2  1177994172_10d143cb8d.jpg#2   
3           3  1177994172_10d143cb8d.jpg#3   
4           4  1177994172_10d143cb8d.jpg#4   
5           5  1232148178_4f45cc3284.jpg#0   
6           6  1232148178_4f45cc3284.jpg#1   
7           7  1232148178_4f45cc3284.jpg#2   
8           8  1232148178_4f45cc3284.jpg#3   
9           9  1232148178_4f45cc3284.jpg#4   

                                          query_text  \
0  Two blonde boys , one in a camouflage shirt an...   
1  Two boys are squirting water guns at each other .   
2            Two boys spraying each other with water   
3  Two children wearing jeans squirt water at eac...   
4  Two young boys are squirting water at each oth...   
5                    A baby girl playing at a park .   
6  A closeup of a child on a playground with adul...   
7  A young boy poses for a picture in front of a ...   
8  A young girl is smiling in front of the camera...   
9  There is a little blond hair girl with a green...   

                       image  
0  1177994172_10d143cb8d.jpg  
1  1177994172_10d143cb8d.jpg  
2  1177994172_10d143cb8d.jpg  
3  1177994172_10d143cb8d.jpg  
4  1177994172_10d143cb8d.jpg  
5  1232148178_4f45cc3284.jpg  
6  1232148178_4f45cc3284.jpg  
7  1232148178_4f45cc3284.jpg  
8  1232148178_4f45cc3284.jpg  
9  1232148178_4f45cc3284.jpg



"df_test" состоит из:
500 строк
4 столбцов


Общая информация о "df_test"

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 500 entries, 0 to 499
Data columns (total 4 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Unnamed: 0  500 non-null    int64 
 1   query_id    500 non-null    object
 2   query_text  500 non-null    object
 3   image       500 non-null    object
dtypes: int64(1), object(3)
memory usage: 15.8+ KB


Описательная статистика "df_test"


count unique                                                top  \
Unnamed: 0  500.0    NaN                                                NaN   
query_id      500    500                         989851184_9ef368e520.jpg#4   
query_text    500    500  The black dog has a toy in its mouth and a per...   
image         500    100                          1177994172_10d143cb8d.jpg   

           freq   mean         std  min     25%    50%     75%    max  
Unnamed: 0  NaN  249.5  144.481833  0.0  124.75  249.5  374.25  499.0  
query_id      1    NaN         NaN  NaN     NaN    NaN     NaN    NaN  
query_text    1    NaN         NaN  NaN     NaN    NaN     NaN    NaN  
image         5    NaN         NaN  NaN     NaN    NaN     NaN    NaN

In [8]:
df_test.sample(15)

Unnamed: 0                     query_id  \
80           80  2308256827_3c0a7d514d.jpg#0   
423         423  3730011701_5352e02286.jpg#3   
92           92  2399219552_bbba0a9a59.jpg#2   
358         358  3466891862_9afde75568.jpg#3   
35           35  1745110280_0cbff5e273.jpg#0   
236         236  3122579598_14841c038a.jpg#1   
489         489   973827791_467d83986e.jpg#4   
272         272  3211199368_ca78387f72.jpg#2   
376         376  3514297698_0512623955.jpg#1   
282         282  3262849619_0bc4f88ef9.jpg#2   
353         353  3459871361_92d1ecda36.jpg#3   
146         146  2670637584_d96efb8afa.jpg#1   
18           18  1319634306_816f21677f.jpg#3   
467         467   782401952_5bc5d3413a.jpg#2   
318         318  3396275223_ee080df8b5.jpg#3   

                                            query_text  \
80   A boy walks across a rope structure on a playg...   
423  The little girl in the yellow hat is buried up...   
92   A little boy plays with a basketball and a toy...   
358     The small dog returns with his stick proudly .   
35   A brown dog with a black collar stands next to...   
236         Two boys and a white dog stand in a park .   
489  I child slides into a wading pool full of wate...   
272  A dog is wearing jeans and a blue and yellow s...   
376  A dog is catching a yellow tennis ball on the ...   
282     A brown and white dog is carrying a snowball .   
353  A young boy jumps his bicycle on a tagged bike...   
146  A child sits in the woods looking up with bino...   
18           A large yellow dog is sitting on a hill .   
467  Four young people with cold drinks stand on th...   
318                     Performers performing a play .   

                         image  
80   2308256827_3c0a7d514d.jpg  
423  3730011701_5352e02286.jpg  
92   2399219552_bbba0a9a59.jpg  
358  3466891862_9afde75568.jpg  
35   1745110280_0cbff5e273.jpg  
236  3122579598_14841c038a.jpg  
489   973827791_467d83986e.jpg  
272  3211199368_ca78387f72.jpg  
376  3514297698_0512623955.jpg  
282  3262849619_0bc4f88ef9.jpg  
353  3459871361_92d1ecda36.jpg  
146  2670637584_d96efb8afa.jpg  
18   1319634306_816f21677f.jpg  
467   782401952_5bc5d3413a.jpg  
318  3396275223_ee080df8b5.jpg

**Вывод по загрузке данных**:
- с таблицей `df_train` всё в порядке
- в таблицах `df_crowdAnnotations` и `df_expertAnnotations` первые записили превратились в названия столбцов (убрал при загрузке таблиц)
- в таблице `df_test` первый столбец повторяет `id`, поэтому его можно удалить

In [9]:
df_test = df_test.drop('Unnamed: 0', axis=1)

In [10]:
df_test.head(5)

query_id  \
0  1177994172_10d143cb8d.jpg#0   
1  1177994172_10d143cb8d.jpg#1   
2  1177994172_10d143cb8d.jpg#2   
3  1177994172_10d143cb8d.jpg#3   
4  1177994172_10d143cb8d.jpg#4   

                                          query_text  \
0  Two blonde boys , one in a camouflage shirt an...   
1  Two boys are squirting water guns at each other .   
2            Two boys spraying each other with water   
3  Two children wearing jeans squirt water at eac...   
4  Two young boys are squirting water at each oth...   

                       image  
0  1177994172_10d143cb8d.jpg  
1  1177994172_10d143cb8d.jpg  
2  1177994172_10d143cb8d.jpg  
3  1177994172_10d143cb8d.jpg  
4  1177994172_10d143cb8d.jpg

In [11]:
for df in df_list:
    print(f'Пропущенные значения "{df.name}":\n{df.isna().sum()}')
    print('\n')

Пропущенные значения "df_train":
image         0
query_id      0
query_text    0
dtype: int64


Пропущенные значения "df_crowdAnnotations":
image         0
query_id      0
desc_perc     0
n_match       0
n_no_match    0
dtype: int64


Пропущенные значения "df_expertAnnotations":
image       0
query_id    0
ex_1        0
ex_2        0
ex_3        0
dtype: int64


Пропущенные значения "df_test":
Unnamed: 0    0
query_id      0
query_text    0
image         0
dtype: int64




In [12]:
for df in df_list:
    print(f'Явные дубликаты "{df.name}":\n{df.duplicated().sum()}')
    print('\n')

Явные дубликаты "df_train":
0


Явные дубликаты "df_crowdAnnotations":
0


Явные дубликаты "df_expertAnnotations":
0


Явные дубликаты "df_test":
0




In [13]:
datagen = ImageDataGenerator(rescale=1./255)
sample_generator = datagen.flow_from_dataframe(
    dataframe=df_train.head(5),
    directory=os.path.join(DATA_PATH, 'train_images'),
    x_col='image',
    y_col='query_text',
    target_size=(224, 224),
    batch_size=15,
    class_mode='raw'
)

Found 5 validated image filenames.


In [14]:
images, queries = next(sample_generator)

In [15]:
num_images = len(images)
plt.figure(figsize=(10, 5 * num_images))

for i in range(num_images):
    plt.subplot(num_images, 1, i+1)
    plt.imshow(images[i])
    plt.title(f"Query: {queries[i]}", fontsize=12)
    plt.axis('off')

plt.tight_layout()
plt.show()

Output hidden; open in https://colab.research.google.com to view.

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Насколько я вижу ты закончил анализ и приступаешь к подготовке данных для обучения. Поэтому давай закочим раздел "Анализ", сформулируем выводы и начнем новый раздел..</div>


<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
<b>v2</b>    
Не исправлено
</div>

**Вывод**:
- дубликатов не обнаружено
- пропусков не обнаружено
- некоторые описания совершенно не подходят картинкам

In [16]:
# Функция для агрегации экспертных оценок
def aggregate_scores(row):
    scores = [row['ex_1'], row['ex_2'], row['ex_3']]

    if len(set(scores)) == 3:
        return None

    modal_score = mode(scores)
    return modal_score[0]

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Когда я проверял твой проект (scipy==1.8.0), то после выполнения этого кода вместо значений появились numpy.array. И дальнейший код ломался.
    
Возможно стоит возвращать modal_score[0][0]. Но здесь все может зависеть от версии scipy.
</div>

In [17]:
# Применяем агрегацию
df_expertAnnotations['expert_score'] = df_expertAnnotations.apply(aggregate_scores, axis=1)
# Удаляем строки, где эксперты сильно разошлись (вернулся None)
df_expertAnnotations = df_expertAnnotations.dropna(subset=['expert_score'])
# Нормализуем оценку в диапазон [0, 1] (делением на 4)
df_expertAnnotations['expert_score_norm'] = (df_expertAnnotations['expert_score'] - 1) / 3.0

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Если ты просто применишь деление на 4, то из диапазона 1-4 мы перейдем в диапазон 0,25-1.  Чтобы получился димпазон 0-1 нужно использовать другое преобразование (x-1)/3</div>


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
    <b>v2</b> исправлено 👍 </div>

In [18]:
df_expertAnnotations.head()

image                     query_id  ex_1  ex_2  ex_3  \
0  1056338697_4f7d7ce270.jpg  2549968784_39bfbe44f9.jpg#2     1     1     1   
1  1056338697_4f7d7ce270.jpg  2718495608_d8533e3ac5.jpg#2     1     1     2   
2  1056338697_4f7d7ce270.jpg  3181701312_70a379ab6e.jpg#2     1     1     2   
3  1056338697_4f7d7ce270.jpg  3207358897_bfa61fa3c6.jpg#2     1     2     2   
4  1056338697_4f7d7ce270.jpg  3286822339_5535af6b93.jpg#2     1     1     2   

   expert_score  expert_score_norm  
0           1.0           0.000000  
1           1.0           0.000000  
2           1.0           0.000000  
3           2.0           0.333333  
4           1.0           0.000000

In [19]:
# Вычисляем общее число голосов
df_crowdAnnotations['total_votes'] = df_crowdAnnotations['n_match'] + df_crowdAnnotations['n_no_match']
# Вычисляем "уверенность" оценки (чем больше голосов, тем увереннее)
df_crowdAnnotations['confidence'] = 1 - 1/(df_crowdAnnotations['total_votes'] + 1)
# Взвешенная оценка
df_crowdAnnotations['crowd_score_norm'] = df_crowdAnnotations['desc_perc'] * df_crowdAnnotations['confidence']

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
немного не согласен с тем, как ты используешь confidence.
    
Она просто занижает оценку и всё.  Высокие оценки станут ниже, а нули так и останутся нулями.
    
Как альтернативу можно рассмотреть исключение оценок, в которых мы не уверены в оценках краудсорсеров.    
</div>

In [20]:
df_crowdAnnotations.head()

image                     query_id  desc_perc  n_match  \
0  1056338697_4f7d7ce270.jpg  1056338697_4f7d7ce270.jpg#2        1.0        3   
1  1056338697_4f7d7ce270.jpg   114051287_dd85625a04.jpg#2        0.0        0   
2  1056338697_4f7d7ce270.jpg  1427391496_ea512cbe7f.jpg#2        0.0        0   
3  1056338697_4f7d7ce270.jpg  2073964624_52da3a0fc4.jpg#2        0.0        0   
4  1056338697_4f7d7ce270.jpg  2083434441_a93bc6306b.jpg#2        0.0        0   

   n_no_match  total_votes  confidence  crowd_score_norm  
0           0            3        0.75              0.75  
1           3            3        0.75              0.00  
2           3            3        0.75              0.00  
3           3            3        0.75              0.00  
4           3            3        0.75              0.00

In [21]:
def merge_annotations(df_expertAnnotations, df_crowdAnnotations):
    merged = pd.merge(
        df_expertAnnotations[['image', 'query_id', 'expert_score_norm']],
        df_crowdAnnotations[['image', 'query_id', 'crowd_score_norm', 'total_votes']],
        on=['image', 'query_id'],
        how='outer'  # Используем outer join для сохранения всех возможных пар
    )

    merged['target'] = np.where(
        merged['expert_score_norm'].notna() & merged['crowd_score_norm'].notna(),
        0.7 * merged['expert_score_norm'] + 0.3 * merged['crowd_score_norm'],
        np.where(
            merged['expert_score_norm'].notna(),
            merged['expert_score_norm'],
            np.where(
                merged['crowd_score_norm'].notna(),
                merged['crowd_score_norm'],
                np.nan
            )
        )
    )

    return merged

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">

По объединению экспертных и крауд оценок все ОК, но подсвечу возможные альтернативы
    
--------------    
    
    
В этом проекте есть несколько возможных стратегий работы с обучающими и краудсорс оценками,  вот эти стратегии с плюсами и минусами:
   

   - использовать соединение через outer, в этом случае, при корректной обработке пропусков мы можем получить более 50К размеченых пар. К плюсам данного подхода можно отнести большое количество данных. Минус - в основном это будут данные, размеченные краудсорсерами, а там качество раметки ниже.
   - использовать только экспертные оценки. плюс - высокое качество данных (разметка имеет шкалу). Минус - данных меньше.  Поясню по поводу качества разметки на примере: если на изображении одна собака, а в описании "две собаки бегут по берегу", то эксперты поставять соответствие 0,6-0,7 (переводя в шкалу 0-1), а краудсорсеры поставят 0.  Но на экспертных оценках модель хотя-бы научится находить собак, а на краудсор оценках ничему не начится.

   - использование соединения через left.  Данный подход очевидно проигрывает второму варианту: данных столько же, нужно дополнительно возиться с объединением, но качество разметки снижается.. Но это соответсвует предложениям авторов проекта, поэтому такой подход имеет право на жизнь

   - объединение через inner. Тут все плохо - и данных ОЧЕНЬ мало (в два раза меньше, чем использовать только экспертные оценки) и качество у них "подпорчено". И навыков по обработке пропусков не получаем.....


</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
   
    
Но заполнять пропуски значениями 0,5 категорически не стоит.
  
В объединенных данных около 5% пар где встречаются и крауд и экспертные оценки. Заполнив недостающие данные ты вносишь сильнейший шум.
    
Как альтернативу советую либо вообще отказаться от крауд оценок (работать с экспертными) либо честно использовать оценки:
    
 - если есть только крауд - использовать их
 - если есть только экспертные - использовать их
 - а там где встречаются обе - использовать взвешенную сумму
    
</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
    <b>v2</b> исправлено 👍 </div>

In [22]:
annotations_merged = merge_annotations(df_expertAnnotations, df_crowdAnnotations)

In [23]:
annotations_merged.head()

image                     query_id  expert_score_norm  \
0  1056338697_4f7d7ce270.jpg  1056338697_4f7d7ce270.jpg#2                NaN   
1  1056338697_4f7d7ce270.jpg   114051287_dd85625a04.jpg#2                NaN   
2  1056338697_4f7d7ce270.jpg  1427391496_ea512cbe7f.jpg#2                NaN   
3  1056338697_4f7d7ce270.jpg  2073964624_52da3a0fc4.jpg#2                NaN   
4  1056338697_4f7d7ce270.jpg  2083434441_a93bc6306b.jpg#2                NaN   

   crowd_score_norm  total_votes  target  
0              0.75          3.0    0.75  
1              0.00          3.0    0.00  
2              0.00          3.0    0.00  
3              0.00          3.0    0.00  
4              0.00          3.0    0.00

In [24]:
annotations_merged.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 51268 entries, 0 to 51267
Data columns (total 6 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   image              51268 non-null  object 
 1   query_id           51268 non-null  object 
 2   expert_score_norm  5696 non-null   float64
 3   crowd_score_norm   47830 non-null  float64
 4   total_votes        47830 non-null  float64
 5   target             51268 non-null  float64
dtypes: float64(4), object(2)
memory usage: 2.3+ MB


In [25]:
annotations_merged.describe()

expert_score_norm  crowd_score_norm   total_votes        target
count        5696.000000      47830.000000  47830.000000  51268.000000
mean            0.203827          0.051686      3.028643      0.060959
std             0.286367          0.155878      0.208075      0.167022
min             0.000000          0.000000      3.000000      0.000000
25%             0.000000          0.000000      3.000000      0.000000
50%             0.000000          0.000000      3.000000      0.000000
75%             0.333333          0.000000      3.000000      0.000000
max             1.000000          0.833333      6.000000      1.000000

In [26]:
annotations_merged['target'].sort_values().unique()

array([0.        , 0.06      , 0.075     , 0.14285714, 0.16666667,
       0.2       , 0.23333333, 0.25      , 0.28333333, 0.28571429,
       0.29333333, 0.30833333, 0.33333333, 0.33333333, 0.35333333,
       0.38333333, 0.4       , 0.45833333, 0.46666667, 0.5       ,
       0.5       , 0.52666667, 0.54166667, 0.55238095, 0.56666667,
       0.57142857, 0.58666667, 0.5952381 , 0.6       , 0.61666667,
       0.63809524, 0.64666667, 0.66666667, 0.66666667, 0.69166667,
       0.7       , 0.75      , 0.775     , 0.8       , 0.83333333,
       0.85      , 0.925     , 0.94      , 0.95      , 1.        ])

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
   
Крайне важно делать самопроверку. Сейчас ты получил датафрейм annotations_merged. Изучи его (как минимум выведи несколко строк + посмотри info)
    
</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
<b>v2</b>    
Не исправлено
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента</b></font>
   
Почему? я же про `annotations_merged` всё вывел выше.

</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
<b>v2</b>    
я очень тебя прошу убрать из проекта весь лишний закомментированный  код.  Проект и так сложный, а если его засорять, то с ним становится крайне сложно работать.  Если он тебе зачем-то  нужен - можешь сохранить его в отдельном черновике
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента</b></font>
   
Случайно захламил проект. Пока исправлял, чтобы не запутаться и ничего не забыть комментировал старый код, делал поправки отталкиваясь от него и забыл удалить

</div>

In [27]:
queries = df_train[['query_id', 'query_text']].drop_duplicates()
queries.shape

(977, 2)

In [28]:
df_train = pd.merge(
    annotations_merged,
    queries,
    on='query_id',
    how='inner'
)

In [29]:
df_train.shape

(50159, 7)

In [30]:
df_train.head()

image                     query_id  expert_score_norm  \
0  1056338697_4f7d7ce270.jpg  1056338697_4f7d7ce270.jpg#2                NaN   
1  1056338697_4f7d7ce270.jpg   114051287_dd85625a04.jpg#2                NaN   
2  1056338697_4f7d7ce270.jpg  1427391496_ea512cbe7f.jpg#2                NaN   
3  1056338697_4f7d7ce270.jpg  2073964624_52da3a0fc4.jpg#2                NaN   
4  1056338697_4f7d7ce270.jpg  2083434441_a93bc6306b.jpg#2                NaN   

   crowd_score_norm  total_votes  target  \
0              0.75          3.0    0.75   
1              0.00          3.0    0.00   
2              0.00          3.0    0.00   
3              0.00          3.0    0.00   
4              0.00          3.0    0.00   

                                          query_text  
0  A woman is signaling is to traffic , as seen f...  
1          A boy in glasses is wearing a red shirt .  
2   A young boy holds onto a blue handle on a pier .  
3  A woman wearing black clothes , a purple scarf...  
4  An older woman with blond hair rides a bicycle...

In [31]:
df_train = df_train.reindex(columns=['image', 'query_id', 'query_text', 'expert_score_norm', 'crowd_score_norm', 'total_votes', 'target'])

In [32]:
df_train.head()

image                     query_id  \
0  1056338697_4f7d7ce270.jpg  1056338697_4f7d7ce270.jpg#2   
1  1056338697_4f7d7ce270.jpg   114051287_dd85625a04.jpg#2   
2  1056338697_4f7d7ce270.jpg  1427391496_ea512cbe7f.jpg#2   
3  1056338697_4f7d7ce270.jpg  2073964624_52da3a0fc4.jpg#2   
4  1056338697_4f7d7ce270.jpg  2083434441_a93bc6306b.jpg#2   

                                          query_text  expert_score_norm  \
0  A woman is signaling is to traffic , as seen f...                NaN   
1          A boy in glasses is wearing a red shirt .                NaN   
2   A young boy holds onto a blue handle on a pier .                NaN   
3  A woman wearing black clothes , a purple scarf...                NaN   
4  An older woman with blond hair rides a bicycle...                NaN   

   crowd_score_norm  total_votes  target  
0              0.75          3.0    0.75  
1              0.00          3.0    0.00  
2              0.00          3.0    0.00  
3              0.00          3.0    0.00  
4              0.00          3.0    0.00

In [33]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50159 entries, 0 to 50158
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   image              50159 non-null  object 
 1   query_id           50159 non-null  object 
 2   query_text         50159 non-null  object 
 3   expert_score_norm  5696 non-null   float64
 4   crowd_score_norm   46721 non-null  float64
 5   total_votes        46721 non-null  float64
 6   target             50159 non-null  float64
dtypes: float64(4), object(3)
memory usage: 2.7+ MB


<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">

А вот логику последних трех ячеек я если честно совсем не понял. Да и полученный результат просто удручает.
    
Ты зачем-то выделяешь 'desc_num', но это просто не имеет смысла, т.к. все описания в обучающих данных заканчиваются на `#2` (этот момент стоило выявить на этапе знакомства с данными).
    
Теперь давай поймем, какую задачу ты хочешь решить? В датафрейме `annotations_merged` есть вся необходимая информация кроме расшифровки текстового описания.
        
Восстановить текстовое описание мы можем следующим образом:
 - создать датафрейм с УНИКАЛЬНЫМИ ПАРАМИ query_id/query_text(их всего 977)
 - присоединить этот датафрейм используя inner/left
 - важно проконтролировать, что у нас останется столько же записей, сколько было в датафрейме `annotations_merged`
    
Но делаешь ты что-то совершенно другое, как следствие вначале датафрейм разразтается до 300 тысяч записей, но после удаления дублей там остается меньше 1000 записей.
    
Более того, если глазами посмотреть на те несколько строк, что ты выводишь ниже, то видно, что для изображения 2549968784_39bfbe44f9.jpg и описания 2549968784_39bfbe44f9.jpg#2 метрика сходства равна 0,35, тогда как это описание создано специально под это изображение, соответственно сходство должно быть идеальным - 1.0
    

</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
    <b>v2</b> исправлено 👍 </div>

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Пожалуйста посмотри info по полученному датафрейму
</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
<b>v2</b>    
Не исправлено
</div>

**Вывод**:
- исправили ошибки в данных
- агрегировали оценки экспертов и краудсорсинговые оценки
- создали `target`

## 2. Проверка данных

В некоторых странах, где работает ваша компания, действуют ограничения по обработке изображений: поисковым сервисам и сервисам, предоставляющим возможность поиска, запрещено без разрешения родителей или законных представителей предоставлять любую информацию, в том числе, но не исключительно тексты, изображения, видео и аудио, содержащие описание, изображение или запись голоса детей. Ребёнком считается любой человек, не достигший 16 лет.

В вашем сервисе строго следуют законам стран, в которых работают. Поэтому при попытке посмотреть изображения, запрещённые законодательством, вместо картинок показывается дисклеймер:

> This image is unavailable in your country in compliance with local laws
>

Однако у вас в PoC нет возможности воспользоваться данным функционалом. Поэтому все изображения, которые нарушают данный закон, нужно удалить из обучающей выборки.

In [34]:
banned_keywords = r'child|kid|baby|babies|boy|girl'

In [35]:
df_train['source_image'] = df_train['query_id'].str.split('#').str[0]
banned_descriptions = df_train[df_train['query_text'].str.contains(banned_keywords, case=False, na=False)]
banned_source_images = banned_descriptions['source_image'].unique()
df_train_filtered = df_train[~df_train['source_image'].isin(banned_source_images)]
df_train_filtered = df_train_filtered.drop(columns=['source_image'])

In [36]:
df_train_filtered.head()

image                     query_id  \
0  1056338697_4f7d7ce270.jpg  1056338697_4f7d7ce270.jpg#2   
3  1056338697_4f7d7ce270.jpg  2073964624_52da3a0fc4.jpg#2   
4  1056338697_4f7d7ce270.jpg  2083434441_a93bc6306b.jpg#2   
5  1056338697_4f7d7ce270.jpg  2204550058_2707d92338.jpg#2   
6  1056338697_4f7d7ce270.jpg  2224450291_4c133fabe8.jpg#2   

                                          query_text  expert_score_norm  \
0  A woman is signaling is to traffic , as seen f...                NaN   
3  A woman wearing black clothes , a purple scarf...                NaN   
4  An older woman with blond hair rides a bicycle...                NaN   
5  A woman wearing a white glove peers over the c...                NaN   
6      A man taking a picture with a silver camera .                NaN   

   crowd_score_norm  total_votes  target  
0              0.75          3.0    0.75  
3              0.00          3.0    0.00  
4              0.00          3.0    0.00  
5              0.00          3.0    0.00  
6              0.00          3.0    0.00

In [37]:
df_train_filtered.info()

<class 'pandas.core.frame.DataFrame'>
Index: 36017 entries, 0 to 50158
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   image              36017 non-null  object 
 1   query_id           36017 non-null  object 
 2   query_text         36017 non-null  object 
 3   expert_score_norm  4178 non-null   float64
 4   crowd_score_norm   33581 non-null  float64
 5   total_votes        33581 non-null  float64
 6   target             36017 non-null  float64
dtypes: float64(4), object(3)
memory usage: 2.2+ MB


<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
    
Ты удаляешь комментарии, в которых есть слова, обозначающие детей. Но наша задача убрать изображения, содержащие детей. Есть два способа сделать это:
    
 - Определить список плохих изображений, как изображения удовлетворяющие условиям:
    - Комментарий содержит плохие слова
    - <b>Оценка соответствия комментария и изображения выше порога.</b>
 - Удалить из нашего датасета все пары "изображение/описание" с плохими изображениями.
    

И второй, очень красивый способ. Он основан на том, что `query_id` содержит в с себе имя изображения, для которого он был написан (такая пасхалочка от авторов датасета).
    
 - Определить список плохих комментариев
 - У `query_id`  плохих комментариев отрезать два последних символа и получим список плохих изображений.
    
    
    
    
    
</div>

**Вывод**
- удалили запрщенные изображения

## 3. Векторизация изображений

Перейдём к векторизации изображений.

Самый примитивный способ — прочесть изображение и превратить полученную матрицу в вектор. Такой способ нам не подходит: длина векторов может быть сильно разной, так как размеры изображений разные. Поэтому стоит обратиться к свёрточным сетям: они позволяют "выделить" главные компоненты изображений. Как это сделать? Нужно выбрать какую-либо архитектуру, например ResNet-18, посмотреть на слои и исключить полносвязные слои, которые отвечают за конечное предсказание. При этом можно загрузить модель данной архитектуры, предварительно натренированную на датасете ImageNet.

In [38]:
# Загружаем модель и убираем последний классификационный слой
resnet = models.resnet18(pretrained=True)
resnet = torch.nn.Sequential(*list(resnet.children())[:-1])  # удаляем последний слой
resnet.eval();  # переводим в режим инференса

Downloading: "https://download.pytorch.org/models/resnet18-f37072fd.pth" to /root/.cache/torch/hub/checkpoints/resnet18-f37072fd.pth
100%|██████████| 44.7M/44.7M [00:00<00:00, 136MB/s]


<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Непонятно, выводишь ли ты архитектуру resnet18 сознательно или это случайный эффект. Если хочешь убрать - просто добавь точку с запятой в конец последнй строки.</div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
<b>минутка теории</b>
 По умолчанию jupyter notebook выводит результат вычислений в последней строке ячейки, если этот результат не записан в переменную. Мы часто пользуемся этим, обратившись к какому-нибудь объекту в последней строке и видим результат без дополнительных print() или dispay()
    
    
  Ниже три ячейки с кодом. Обрати внимание, если просто сделать вычисление - то результат выводится, но если добавить точку с запятой или записать результат в переменную, то ничего не происходит. Обрати внимание, когда происходит вывод:
   - последня строка
   - нет точки с запятой в конце
    
   - результат не записан в переменную

И досточно убрать одно из этих условий, как вывода нет.

![image-4.png](attachment:image-4.png)    

 Функции matlitlib не только выводят графики, но и возвращают python объекты axis. Это внутренняя кухня, чтобы можно было строить сложные графики. Но для нас побочным эффектом является то, что в ноутбуке мы видим текстовое предоставление объекта, которое выводит ноутбук. И для красоты такое поведение стоит заблокировать.



</div>

In [39]:
# Трансформации такие же, как при обучении на ImageNet
transform = transforms.Compose([
    transforms.Resize((224, 224)),  # Resize до нужного размера
    transforms.ToTensor(),
    transforms.Normalize(mean=[0.485, 0.456, 0.406],
                         std=[0.229, 0.224, 0.225])
])

In [40]:
def get_image_embedding(img_path, model, transform):
    img = Image.open(img_path).convert('RGB')
    img_t = transform(img).unsqueeze(0)  # добавим размер батча
    with torch.no_grad():
        embedding = model(img_t).squeeze().numpy()
    return embedding

In [41]:
image_embeddings = {}

for image_name in tqdm(os.listdir(image_folder)):
    img_path = os.path.join(image_folder, image_name)
    try:
        emb = get_image_embedding(img_path, resnet, transform)
        image_embeddings[image_name] = emb
    except Exception as e:
        print(f'Ошибка с изображением {image_name}: {e}')

100%|██████████| 1000/1000 [01:49<00:00,  9.11it/s]


**Вывод**
- сделали векторизацию изображений с помощью `resnet18`

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
 Отлично, вектра для изображений получены👍 </div>

## 4. Векторизация текстов

Следующий этап — векторизация текстов. Вы можете поэкспериментировать с несколькими способами векторизации текстов:

- tf-idf
- word2vec
- \*трансформеры (например Bert)

\* — если вы изучали трансформеры в спринте Машинное обучение для текстов.


In [42]:
# Загружаем BERT-модель и токенизатор
tokenizer = BertTokenizer.from_pretrained('bert-base-uncased', force_download=True)
bert_model = BertModel.from_pretrained('bert-base-uncased')
bert_model.eval();  # переводим в режим инференса

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/570 [00:00<?, ?B/s]

Xet Storage is enabled for this repo, but the 'hf_xet' package is not installed. Falling back to regular HTTP download. For better performance, install the package with: `pip install huggingface_hub[hf_xet]` or `pip install hf_xet`


model.safetensors:   0%|          | 0.00/440M [00:00<?, ?B/s]

In [43]:
def get_text_embedding(text, tokenizer, model):
    tokens = tokenizer(text, return_tensors='pt', truncation=True, padding=True, max_length=64)
    with torch.no_grad():
        output = model(**tokens)
    # Получаем эмбеддинги токенов и делаем mean pooling
    embeddings = output.last_hidden_state.squeeze(0)
    attention_mask = tokens['attention_mask'].squeeze(0)
    mask_expanded = attention_mask.unsqueeze(-1).expand(embeddings.size()).float()
    mean_pooled = torch.sum(embeddings * mask_expanded, dim=0) / torch.clamp(mask_expanded.sum(dim=0), min=1e-9)
    return mean_pooled.numpy()

In [44]:
text_embeddings = {}

for idx, row in tqdm(df_train_filtered.iterrows(), total=len(df_train_filtered)):
    text = row['query_text']
    query_id = row['query_id']

    if query_id not in text_embeddings:
        try:
            emb = get_text_embedding(text, tokenizer, bert_model)
            text_embeddings[query_id] = emb
        except Exception as e:
            print(f'Ошибка с query_id={query_id}: {e}')


100%|██████████| 36017/36017 [01:17<00:00, 461.85it/s]  


<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Отлично,полученные векторные представления для текстов описаний с помощью BERT </div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
Очень круто, что ты позаботился об оптимизации вычислений, и векторизуешь только уникальные тексты.

Похожего эффекта можно достичь если использовать стандартный декоратор lru_cache (можешь посмотреть примеры, легко гуглится)
</div>

**вывод**
- сделали векторизацию описаний с помощью `Bert`

## 5. Объединение векторов

Подготовьте данные для обучения: объедините векторы изображений и векторы текстов с целевой переменной.

In [45]:
df_train_filtered_unique = df_train_filtered.drop_duplicates(subset=['image', 'query_id'])
print(len(df_train_filtered), "->", len(df_train_filtered_unique))

36017 -> 36017


<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Как раз то, о чем я писал тебе выше. Из-за некорректной подготовки обучающего датасета у нас осталось всего 698 пар для обучения модели что явно маловато. </div>

<img src="https://emojigraph.org/media/apple/check-mark-button_2705.png" align=left width=33, heigth=33>
<div class="alert alert-success">
    <b>v2</b> исправлено 👍 </div>

In [46]:
X = []
y = []

for idx, row in tqdm(df_train_filtered_unique.iterrows(), total=len(df_train_filtered_unique)):
    image_name = row['image']
    query_id = row['query_id']
    target = row['target']

    if image_name in image_embeddings and query_id in text_embeddings:
        image_emb = image_embeddings[image_name]
        text_emb = text_embeddings[query_id]
        combined = np.concatenate([image_emb, text_emb])
        X.append(combined)
        y.append(target)


100%|██████████| 36017/36017 [00:02<00:00, 12835.08it/s]


In [47]:
X = np.array(X)
y = np.array(y)

In [48]:
X.shape
y.shape

(36017,)

**Вывод**
- Объединили векторы изображений и описаний для обучения моделей

## 6. Обучение модели предсказания соответствия

Для обучения разделите датасет на тренировочную и тестовую выборки. Простое случайное разбиение не подходит: нужно исключить попадание изображения и в обучающую, и в тестовую выборки.
Для того чтобы учесть изображения при разбиении, можно воспользоваться классом [GroupShuffleSplit](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GroupShuffleSplit.html) из библиотеки sklearn.model_selection.

Код ниже разбивает датасет на тренировочную и тестовую выборки в пропорции 7:3 так, что строки с одинаковым значением 'group_column' будут содержаться либо в тестовом, либо в тренировочном датасете.

```
from sklearn.model_selection import GroupShuffleSplit
gss = GroupShuffleSplit(n_splits=1, train_size=.7, random_state=42)
train_indices, test_indices = next(gss.split(X=df.drop(columns=['target']), y=df['target'], groups=df['group_column']))
train_df, test_df = df.loc[train_indices], df.loc[test_indices]

```

Какую модель использовать — выберите самостоятельно. Также вам предстоит выбрать метрику качества либо реализовать свою.

In [49]:
# Мы будем группировать по названию изображения
groups = df_train_filtered_unique['image'].values

# Разделим на 70% для тренировки и 30% для теста
gss = GroupShuffleSplit(n_splits=1, train_size=0.7, random_state=42)

# Разбиваем индексы на train и valid
train_indices, valid_indices = next(gss.split(X, y, groups=groups))

# Получаем тренировочный и тестовый датасеты
X_train, X_valid = X[train_indices], X[valid_indices]
y_train, y_valid = y[train_indices], y[valid_indices]

print(f"Размер обучающего набора: {X_train.shape}")
print(f"Размер тестового набора: {X_valid.shape}")

Размер обучающего набора: (25055, 1280)
Размер тестового набора: (10962, 1280)


In [50]:
# Создаём пайплайн для линейной регрессии с масштабированием
pipeline_lr = make_pipeline(StandardScaler(), LinearRegression())

# Параметры для поиска
param_grid_lr = {
    'linearregression__fit_intercept': [True, False]
}

# Настроим GridSearchCV
grid_lr = GridSearchCV(pipeline_lr, param_grid_lr, cv=5, n_jobs=-1)
grid_lr.fit(X_train, y_train)

# Лучшая модель
best_lr_model = grid_lr.best_estimator_

# Прогнозируем с лучшей моделью
y_pred_best_lr = best_lr_model.predict(X_valid)

# Метрики качества
mae_best_lr = mean_absolute_error(y_valid, y_pred_best_lr)
r2_best_lr = r2_score(y_valid, y_pred_best_lr)

print(f'Лучшая линейная регрессия - MAE: {mae_best_lr}, R2: {r2_best_lr}')

Лучшая линейная регрессия - MAE: 0.11598510771954608, R2: -0.05576835911727751


In [51]:
X_sample, y_sample = resample(X_train, y_train, n_samples=5000)

In [55]:
# Параметры для поиска
param_grid_mlp = {
    'mlpregressor__hidden_layer_sizes': [(128,64), (256,128)],
    'mlpregressor__activation': ['relu'],
    'mlpregressor__solver': ['adam'],
    'mlpregressor__max_iter': [500]
}

# Настроим GridSearchCV
grid_mlp = GridSearchCV(make_pipeline(StandardScaler(), MLPRegressor()), param_grid_mlp, cv=3, n_jobs=-1)
grid_mlp.fit(X_sample, y_sample)

# Лучшая модель
best_mlp_model = grid_mlp.best_estimator_

# Прогнозируем с лучшей моделью
y_pred_best_mlp = best_mlp_model.predict(X_valid)

# Метрики качества
mae_best_mlp = mean_absolute_error(y_valid, y_pred_best_mlp)
r2_best_mlp = r2_score(y_valid, y_pred_best_mlp)

print(f'Лучшая нейронная сеть - MAE: {mae_best_mlp}, R2: {r2_best_mlp}')

Лучшая нейронная сеть - MAE: 0.21250727591241833, R2: -1.5013231314268798


**Вывод**
- Обучили линейную регрессию, получив метрики `MAE`: 0.11598510771954608 и `R2`: -0.05576835911727751
- Обучили нейронную сеть `MLPRegressor`, получив метрики `MAE`: 0.21250727591241833 и `R2`: -1.5013231314268798
- Хоть у линейной регрессии метрика получилась лучше, будем использовать `MLPRegressor`, потому что нейронные сети лучше работают с изображениями

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
По моделям вроде бы все соответствует тому, как мы работали в других проектах. Но для текущей задачи выбор инструментария мне кажется неподходящим.
    
Начнем определения выборок. Финальное тестирование (визуальное) мы делаем на отложенной выборке. Поэтому выборка,которую ты сейчас отделил играет роль валидационной. Есть смысл в её названии использовать слово valid а не test.
    
Второй момент - использование нейронной сети. Конечно формально MLPRegressor это нейронная сеть. Но в учебном проекте стоит построить и обучить нейронную сеть используя один из DL фреймворков, и если ты уж начал работать с torch (векторизация изображений сделана с помощью torch, также ты использовал torch версию модели bert для векторизации текстов), то логично было бы и итоговую модель обучить исползьуя torch. Это позволит тебе тоньше поработать с архитектурой, использвать батчнормализацию, дропаут и другие продвинутые инструменты.
    
Из первых двух пунктов вытекает отказ от кросс-валидации. Конечно технически можно использовать кросс-валидацию и для torch модели, однако это не позволит тебе вдумчиво поэкспериментировать с разными архитектурами в ручном режиме подбирая разные гиперпараметры модели и по разному настраивая обучение.
    
ну и формально - если ты используешь кросс-валидацию, то вместо стандартной кросс-валидации лучше использовать GroupShuffleSplit.    
    
    
    
</div>

## 7. Тестирование модели

Настало время протестировать модель. Для этого получите эмбеддинги для всех тестовых изображений из папки `test_images`, выберите случайные 10 запросов из файла `test_queries.csv` и для каждого запроса выведите наиболее релевантное изображение. Сравните визуально качество поиска.

In [56]:
df_test['source_image'] = df_test['query_id'].str.split('#').str[0]
banned_descriptions = df_test[df_test['query_text'].str.contains(banned_keywords, case=False, na=False)]
banned_source_images = banned_descriptions['source_image'].unique()
df_test_filtered = df_test[~df_test['source_image'].isin(banned_source_images)]
df_test_filtered = df_test_filtered.drop(columns=['source_image'])

In [57]:
image_embeddings = {}

for image_name in tqdm(os.listdir(test_images)):
    img_path = os.path.join(test_images, image_name)
    try:
        emb = get_image_embedding(img_path, resnet, transform)
        image_embeddings[image_name] = emb
    except Exception as e:
        print(f'Ошибка с изображением {image_name}: {e}')

 90%|█████████ | 91/101 [00:10<00:01,  8.99it/s]

Ошибка с изображением .DS_Store: cannot identify image file '/content/to_upload/test_images/.DS_Store'


100%|██████████| 101/101 [00:11<00:00,  8.88it/s]


In [58]:
def search_image_by_text(
    text_query,
    image_embeddings,
    model,
    tokenizer,
    bert_model,
    top_k: int = 3
) -> dict:
    try:
        text_embedding = get_text_embedding(text_query, tokenizer, bert_model)
        if text_embedding is None:
            return {"error": "Ошибка векторизации текста."}
    except Exception as e:
        return {"error": f"Ошибка обработки текста: {str(e)}"}

    predictions = []
    for img_path, img_emb in image_embeddings.items():
        try:
            combined = np.concatenate([img_emb, text_embedding])
            score = model.predict(combined.reshape(1, -1))[0]
            predictions.append((img_path, float(score)))
        except Exception as e:
            print(f"Ошибка с изображением {img_path}: {str(e)}")
            continue

    if not predictions:
        return {"error": "Не удалось сравнить запрос с изображениями."}

    predictions.sort(key=lambda x: x[1], reverse=True)

    return {
        "query": text_query,
        "top_matches": predictions[:top_k]
    }

In [59]:
def test_search_system(queries, image_embeddings, model, tokenizer, bert_model, image_folder):
    for query in queries.sample(10)['query_text']:
        result = search_image_by_text(
            text_query=query,
            image_embeddings=image_embeddings,
            model=model,
            tokenizer=tokenizer,
            bert_model=bert_model
        )

        print(f"\nТоп-3 результата для запроса: '{query}'")
        for i, (img_path, score) in enumerate(result['top_matches'], 1):
            full_img_path = os.path.join(image_folder, img_path)
            print(f"{i}. {img_path} (score: {score:.4f})")

            try:
                img = Image.open(full_img_path)
                plt.figure(figsize=(5,5))
                plt.imshow(img)
                plt.title(f"Score: {score:.4f}")
                plt.axis('off')
                plt.show()
            except Exception as e:
                print(f"Не удалось загрузить {full_img_path}: {str(e)}")

In [61]:
test_search_system(
    queries=df_test_filtered,
    image_embeddings=image_embeddings,
    model=best_mlp_model,
    tokenizer=tokenizer,
    bert_model=bert_model,
    image_folder=test_images
)

Output hidden; open in https://colab.research.google.com to view.

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
<b>v2</b>    

Судя по характерному набору признаков (идеальная структура и стиль,  type hinting, подробная обработка ошибок с сообщениями, избыточные комментарии, не работает) код был написан с помощью AI-ассистента.
    
Это ни хорошо, ни плохо, но крайне важно понимать, что происходит в коде и уметь делать его отладку, находить ошибки. Т.е. использовать AI-ассистента для решения задач, которые ты делаешь хорошо, можешь быстро проверить (рутины) - это ОК. Но когда модель что-то написла а ты не можешь заставить это работать - это не очень круто, т.к. при этом ты  не  растешь как разрабочик.
    
В твоем случае ошибка достаточно примитивна. Вот в фукнции test_search_system есть строка:
    
```python
img = Image.open(img_path)   
```    
Ты не указал папку, в которой искать изображение, а значит она ищется в какой-то дефолтной папке (домашней директории), не находит и ты получаешь ошибку. Нужно собрать полный путь, используя путь к директории с тестовыми изображениями.   
    
    
    
    
</div>

<div class="alert" style="background-color:#ead7f7;color:#8737bf">
    <font size="3"><b>Комментарий студента</b></font>
   
Опять ничего не получилось( Очень тяжело даётся этот проект на этом моменте

</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
Да, наверное можно и так отфильтровать, но я бы советовал внести контроль запросов внутрь функции по поиску изображений.   Т.е. если функция получает запрос в котором мы находим нежелательные слова,  то вместо изображения возвращается дисклеймер.</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
    
Судя по комментарию `выберите best_lr_model или best_mlp_model` код был сгенерен AI асситентом.  Насколько мне известно в Практикуме нет однозначного запрета на использование моделей для генерации кода, однако перед публикацией в портфолио такие вещи нужно вычистить.</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Здесь я попрошу тебя немного переделать логику инференса.
    
Ты уже на финишной прямой курса и во всю ипользуешь функции. Поэтому я попрошу тебя оформить тестирование следующим образом: нужно написать функцию, способную искать изображение по заданному описанию.

  Что должна делаь функция
    
   - Получать текст
   - Возвращать картинку и метрику близости.
   - Выдавать дисклеймер если запрос некорректный
    
А затем нужно 10 раз применить функцию к 10 случайным запросам.   
    
Т.е. функция должна работать с любым запросом. Поэтому векторизацию запроса мы делаем в теле функции. А вот векторизацию изображений лучше сделать заранее.
    
Внутри функции должны выполняться векторизация текста + получение 100 пар (вектор текста + вектор изображения). Для каждой пары получаем предсказание модели и выбираем пару с наивысшей оценкой соотвтствия.
    
-----------------
    
Фактически ты все это уже сделал. Нужно только немного перестроить логику.    
    
.</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/b/ba/Warning_sign_4.0.png" align=left width=44, heigth=33>
<div class="alert alert-warning">
К сожалению качество поиска сильно хромает, но причины тут достаточно прозрачны - еще на этапе подготовке данных для обучения были допущены ошибки, как следствие модель училась "на мусоре".
  
</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
Пожалуйста не забудь оформить выводы.</div>

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/8/81/Stop_sign.png/240px-Stop_sign.png" align=left width=35, heigth=35>
<div class="alert alert-danger">
<b>v2</b>    
Не исправлено
</div>

**Вывод**
- Модель пытается выдавать изображения, которые подходят описаниям, но получается у неё это плохо
- Возможные улучшения:
 - Выбрать другую нейронную сеть
 - Лучше подобрать параметры обучения
 - Взять больше данных для обучения

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Заключительный комментарий</b>

Подведем итоги. Я вижу, что в целом с проектом ты справляешься.
    
Проведено исследование и предобработка данных. Выполнена векторизация изображений и текстовых описаний. Обучены модели для предсказания степени соответствия, а также реализована логика для инференса и с её помощью проведено тестирование.

    
Пожалуйста отработай все критичные замечания, которые я подсветил красными комментариями.
    
    
Желтые комментарии на твое усмотрение. Но учитывая что наш подход "не взлетел" и нужно хоть что-то показать заказчику, попробуй показать возможный инференс с помщью CLIP. С библиотекой sentence-transformers это делается буквально в 10 строк.    
Жду твою работу на повторное ревью.
</div>

<img src="http://s3.amazonaws.com/pix.iemoji.com/images/emoji/apple/ios-12/256/waving-hand.png" align=left width=44, heigth=44>
<div class="alert alert-info">
<b> Идеи и альтернативные подходы</b>
  Если хочешь получить хорошо работающее решение без обучения, то можешь провести эксперимент с нейронной сетью CLIP от OPEN AI (это не обязательно, только если хочешь добиться хороших результатов в этом проекте).

  Про эту нейронку много где написано, но можешь посмотреть например вот здесь  https://habr.com/ru/post/539312/.

  А для написания  реализации наиболее удобной лично для меня оказалась библиотека sbert. https://www.sbert.net/examples/applications/image-search/README.html  
    
    
<hr>    
    
Также неплохо работает следующий подход без предобученных моделей (к тому же легко реализуемый).
    
    
Вместо того, чтобы подавать на вход модели изображение + текст и получать меру соответствия, поступаем по другому.
    
На вход модели подаем вектор описания, а на выходе учимся предсказывать вектор изображения.
    
По поводу метрик тут ситуация следующая. Точно работает следующий подход
    
  - берем метрику mse, но тогда нужно чтобы модель её уменьшала. А значит в обучении используем только пары заведомо сответствующие друг другу

Можно попробовать использовать косинусное расстояние, она позволяет использовать все данные (если есть соответствие то выдает 1, если нет то 0).
    
Также можно попробовать усложнить модель, например и изображения и тексты отправлять в пространство размерностью 256 и уже в этом пространстве учиться их "сближать".
    
По поводу методов векторизации: я бы использовать resnet18 для изображений (размерность эмбеддинга избражения 512) и БЕРТ для текстов (768).
    
   
</div>

## 8. Выводы

- [x]  Jupyter Notebook открыт
- [ ]  Весь код выполняется без ошибок
- [ ]  Ячейки с кодом расположены в порядке исполнения
- [ ]  Исследовательский анализ данных выполнен
- [ ]  Проверены экспертные оценки и краудсорсинговые оценки
- [ ]  Из датасета исключены те объекты, которые выходят за рамки юридических ограничений
- [ ]  Изображения векторизованы
- [ ]  Текстовые запросы векторизованы
- [ ]  Данные корректно разбиты на тренировочную и тестовую выборки
- [ ]  Предложена метрика качества работы модели
- [ ]  Предложена модель схожести изображений и текстового запроса
- [ ]  Модель обучена
- [ ]  По итогам обучения модели сделаны выводы
- [ ]  Проведено тестирование работы модели
- [ ]  По итогам тестирования визуально сравнили качество поиска